addons/addon-base-raas/packages/base-raas-services/lib/environment/service-catalog/environment-sc-cidr-service.js

```diff
@@ -0,0 +1,221 @@
+  }
+
+  getIpPermission(updateRule, cidrList) {
+    const ipv4List = _.filter(cidrList, cidr => IsCidr.v4(cidr));
+    // const ipv6List = _.filter(cidrList, cidr => IsCidr.v6(cidr));
+
+    const IpPermission = {
+      FromPort: updateRule.fromPort,
+      IpProtocol: updateRule.protocol,
+      IpRanges: _.map(ipv4List, cidr => ({
+        CidrIp: cidr,
+        // For future: Custom description doesn't work, since that has to match exactly what's there in the SG.
+        // This would require storing cidr field in DB
+        // Description: `Updated via SWB`,
+      })),
+      // TODO: Ipv6 calls have to be done separately and cannot habe
+      // Ipv6Ranges: _.map(ipv6List, cidr => ({
+      //   CidrIp: cidr,
+      //   Description: `Updated via SWB`,
+      // })),
+      ToPort: updateRule.toPort,
+    };
+    return IpPermission;
+  }
+
+  // This method is smart enough to determine which IPs to revoke/grant access
+  // by peeking into the ingress rules of the workspace's security group
+  async manageIngressRules(
+    requestContext,
+    { existingEnvironment, securityGroupId, currentIngressRules, updateRequest } = {},
+  ) {
+    const addIpPermissions = [];
+    const revokeIpPermissions = [];
+    const newCidrList = [];
+
+    _.forEach(currentIngressRules, existingRule => {
+      const matchingRule = _.find(updateRequest, {
+        fromPort: existingRule.fromPort,
+        toPort: existingRule.toPort,
+        protocol: existingRule.protocol,
+      });
+      if (matchingRule) {
+        const addCidrs = _.difference(matchingRule.cidrBlocks, existingRule.cidrBlocks);
+        const revokeCidrs = _.difference(existingRule.cidrBlocks, matchingRule.cidrBlocks);
+
+        if (!_.isEmpty(addCidrs)) addIpPermissions.push(this.getIpPermission(matchingRule, addCidrs));
+        if (!_.isEmpty(revokeCidrs)) revokeIpPermissions.push(this.getIpPermission(matchingRule, revokeCidrs));
+        newCidrList.push(matchingRule);
+      }
+    });
+
+    // Get params ready
+    const revokeParams = { GroupId: securityGroupId, IpPermissions: revokeIpPermissions };
+    const authorizeParams = { GroupId: securityGroupId, IpPermissions: addIpPermissions };
+
+    // Start making security group updates
+    const ec2Client = await this.getEc2Client(requestContext, existingEnvironment.id);
+    if (!_.isEmpty(revokeIpPermissions)) await this.revokeSecurityGroupIngress(ec2Client, revokeParams);
+    if (!_.isEmpty(addIpPermissions)) await this.authorizeSecurityGroupIngress(ec2Client, authorizeParams);
+
+    return newCidrList;
+  }
+
+  async revokeSecurityGroupIngress(ec2Client, revokeParams) {
+    try {
+      await ec2Client.revokeSecurityGroupIngress(revokeParams).promise();
+    } catch (err) {
+      // Skip errors that pertain to rule not found
+      if (err.code !== 'EntityNotFound')
+        throw this.boom.internalError(`There was a problem revoking access in security group. Error: ${err}`, true);
+    }
```

Comment:
According to EC2 documentation, security group ingress rule changes follow the eventual consistency method. This special condition is to cover unlikely scenarios when a newly added CIDR block hasn't been updated yet on the SG and is attempted to revoke.

Reply:
In case we get an EntityNotFound errror, are we sure that our request went through? Is there a way to confirm that with a test or is that documented?

Reply:
The request will fail whenever it encounters these errors. We're not ignoring these exceptions, we still throw them (I see now the comment is misleading, I'll change that). The exception catching is just to provide a security group-specific message, instead of showing "Something went wrong", which isn't useful at all.

After a failed/partial CIDR block update, the user would still be able to see the error message and the up to date ingress rules on that workspace.

Reply:
So we are catching the exception here and for any error except EntityNotFound we are throwing the error but for EntityNotFound we seem to be swallowing the exception and moving on since we are in catch block. Am I reading it wrong?

Reply:
You're right. Removing exception swallowing part.